tools/cldr-apps/src/main/java/org/unicode/cldr/web/api/VotingCompletion.java

```diff
@@ -0,0 +1,197 @@
+     * satisfying the criteria, this function took 443 ms when called the first time, and 38 ms when called
+     * again shortly thereafter.
+     *
+     * The count could be cached (per user, etc.) and updated more quickly just when the user votes; that could
+     * be done on either the front or the back end. Probably the best performance gain would be by caching on
+     * the front end.
+     */
+    private synchronized int getVotedPathCount(int userId, String localeId, Level coverageLevel) throws SQLException {
+        final String sql = "SELECT xpath FROM " + DBUtils.Table.VOTE_VALUE
+            + " WHERE submitter=" + userId
+            + " AND locale='" + localeId + "'"
+            + " AND value IS NOT NULL";
+        Connection conn = null;
+        PreparedStatement ps = null;
+        ResultSet rs = null;
+        int votedPathCount = 0;
+        try {
+            conn = sm.dbUtils.getAConnection();
+            ps = DBUtils.prepareStatement(conn, "votingCompletion", sql);
+            rs = ps.executeQuery();
+            if (rs == null) {
+                return 0;
+            }
+            while (rs.next()) {
+                int xpathId = rs.getInt(1);
+                String xpath = sm.xpt.getById(xpathId);
+                Level pathCovLevel = supplementalDataInfo.getCoverageLevel(xpath, localeId);
+                if (pathCovLevel.compareTo(coverageLevel) <= 0) {
+                    ++votedPathCount;
+                }
+            }
+        } finally {
+            DBUtils.close(conn, ps, rs);
+        }
+        return votedPathCount;
+    }
+
+    /**
+     * How many paths exist in the given locale and coverage level?
+     *
+     * @param localeId
+     * @param coverageLevel
+     * @return the number of paths satisfying the criteria
+     *
+     * Possibly worth caching by locale and level; the number never (?) changes between
+     * Survey Tool initialization and shutdown.
+     *
+     * Possibly there are other functions that loop through a CLDRFile during initialization
+     * and they could calculate the totals less expensively, as a side effect.
+     *
+     * Performance as measured on localhost: for fr/comprehensive, this function took
+     * 5735 milliseconds (!) when called the first time, then 30 milliseconds if called
+     * again shortly thereafter. However, that 5735 milliseconds evidently involves
+     * server/locale initialization that only happens once per Survey Tool startup,
+     * and would happen anyway if this function weren't the first thing called.
+     * If the user has already loaded the Dashboard, then this function takes 358 ms
+     * the first time called, then 25 ms if called again shortly thereafter.
+     */
+    private synchronized int getTotalPathCount(String localeId, Level coverageLevel) {
+        final CLDRFile cldrFile = sm.getSTFactory().make(localeId, true, true);
+        int total = 0;
+        for (String xpath : cldrFile.fullIterable()) {
```

Comment:
create a path header factory ... pathHeader = factory.fromPath?
pathHeader.section == SUPPRESS?

Reply:
there could be votes for a suppressed item if a change has happened (though not from old votes?)

Reply:
enum is called SPECIAL
there is getShowRowAction

Reply:
PathHeader.txt
//ldml/dates/calendars/calendar[@type="%A"]/eras/eraAbbr/era[@type="%A"]   ; Special  ; Suppress ; &calendar($1) ; &calField(Eras:abbreviated:none)-$2 ; HIDE

Reply:
skip if "Suppress" or "Special" or "HIDE"
-- but there's an api for that

Reply:
Per email:

Here is the test for missing:

MissingStatus missingStatus = VettingViewer.getMissingStatus(file, path, latin);

To get at the PathHeader "filter" for the paths, use

PathHeader:

    public SurveyToolStatus getSurveyToolStatus() {

The meanings are in 

PathHeader.SurveyToolStatus

The phase can also have an effect, although I think we can probably ignore the tweaks that it makes. But we can compare to "results on the screen" and then see if we need to take it into consideration. That would be in:

CheckCLDR.Phase
CheckCLDR.StatusAction